### ISSUE TITLE
🐛 [firebase_messaging] App crashes when receiving message under certain circumstances

### ISSUE THREAD

**Issue by lucianoiam** [FirebaseExtended/flutterfire#4857]
## Bug report

App crashes when trying to receive a FCM message after the app has been killed using Android's task manager.
The message reaches the device but the app seems unable to restart. Sending a message to the app in the "killed" state generates the following lines in logcat:
```
01-29 12:32:24.372  4755  4755 E AndroidRuntime: FATAL EXCEPTION: main
01-29 12:32:24.372  4755  4755 E AndroidRuntime: Process: XXXXXXXX, PID: 4755
01-29 12:32:24.372  4755  4755 E AndroidRuntime: java.lang.RuntimeException: Unable to create service io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService: java.lang.IllegalStateException: ensureInitializationComplete must be called after startInitialization
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4198)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.app.ActivityThread.access$1500(ActivityThread.java:237)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1932)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7660)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: Caused by: java.lang.IllegalStateException: ensureInitializationComplete must be called after startInitialization
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at io.flutter.embedding.engine.loader.FlutterLoader.ensureInitializationComplete(FlutterLoader.java:193)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at io.flutter.view.FlutterMain.ensureInitializationComplete(FlutterMain.java:79)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService.onCreate(FlutterFirebaseMessagingService.java:71)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4186)
01-29 12:32:24.372  4755  4755 E AndroidRuntime: 	... 8 more
```

### Steps to reproduce

1. Check notifications are properly working while the app is running
2. Kill the app using the swipe up gesture in the task list
3. Send a FCM to the app

### Expected behavior

This should not appear in logcat:

`java.lang.RuntimeException: Unable to create service io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService: java.lang.IllegalStateException: ensureInitializationComplete must be called after startInitialization`

_...or better, this is what probably everyone wants from a notifications plugin ..._

Notifications should appear regardless the app is visible or not in the task list.

## Additional context

firebase_messaging version was 7.0.3 by the time of this report

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, 1.25.0-8.1.pre, on macOS 11.1 20C69 darwin-x64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.50.1)
[✓] Connected device (2 available)
    ! Error: iPad XXXXX is busy: Waiting for Device. Xcode will continue when iPad XXXXX
       is finished. (code -10)

• No issues found!
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>
```

Dart SDK 2.12.0-133.2.beta
Flutter SDK 1.25.0-8.1.pre
flutter_webapp 1.0.3+6

dependencies:
- cupertino_icons 0.1.3
- firebase_messaging 7.0.3 [meta flutter firebase_core]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- flutter_local_notifications 3.0.2 [flutter platform flutter_local_notifications_platform_interface timezone]
- flutter_webview_plugin 0.3.11 [flutter]
- geolocator 6.1.13 [flutter geolocator_platform_interface]
- share 0.6.5+4 [meta mime flutter]
- yaml 2.2.1 [charcode collection string_scanner source_span]

dev dependencies:
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]

transitive dependencies:
- async 2.5.0-nullsafety.3 [collection]
- boolean_selector 2.1.0-nullsafety.3 [source_span string_scanner]
- characters 1.1.0-nullsafety.5
- charcode 1.2.0-nullsafety.3
- clock 1.1.0-nullsafety.3
- collection 1.15.0-nullsafety.5
- fake_async 1.2.0-nullsafety.3 [clock collection]
- firebase_core 0.5.3 [firebase_core_platform_interface flutter quiver meta firebase_core_web]
- firebase_core_platform_interface 2.1.0 [flutter meta plugin_platform_interface quiver]
- firebase_core_web 0.2.1+1 [firebase_core_platform_interface flutter flutter_web_plugins meta js]
- flutter_local_notifications_platform_interface 2.0.0+1 [flutter plugin_platform_interface]
- flutter_web_plugins 0.0.0 [flutter js characters collection meta typed_data vector_math]
- geolocator_platform_interface 1.0.8 [flutter meta plugin_platform_interface vector_math]
- js 0.6.3-nullsafety.3
- matcher 0.12.10-nullsafety.3 [stack_trace]
- meta 1.3.0-nullsafety.6
- mime 0.9.7
- path 1.8.0-nullsafety.3
- platform 2.2.1
- plugin_platform_interface 1.0.3 [meta]
- quiver 2.1.5 [matcher meta]
- sky_engine 0.0.99
- source_span 1.8.0-nullsafety.4 [charcode collection path term_glyph]
- stack_trace 1.10.0-nullsafety.6 [path]
- stream_channel 2.1.0-nullsafety.3 [async]
- string_scanner 1.1.0-nullsafety.3 [charcode source_span]
- term_glyph 1.2.0-nullsafety.3
- test_api 0.2.19-nullsafety.6 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- timezone 0.5.9 [path]
- typed_data 1.3.0-nullsafety.5 [collection]
- vector_math 2.1.0-nullsafety.5

```
</details>

---

**Comment by kari538** [comment]
I agree! A message should reach, whether the app is open or not!

However, it does say this on the package's readme page:

"Data Message on Android

While app is active:
onMessage

When app is running in background:
onMessage while app stays in the background.

While app is terminated:
not supported by plugin, message is lost"

Let's hope they find a fix for it!... In the meantime, you should at least be able to send Notifications even when the app is terminated.

https://pub.dev/packages/firebase_messaging

**Comment by lucianoiam** [comment]
The plugin does seem to receive messages even after manually killing the app and crashes immediately. If receiving messages in such state is not supported, incoming messages should not trigger a crash as that could have unintended side effects in client apps.

Being able to receive messages independently of app state deserves a separate feature request.

**Comment by kari538** [comment]
...But what do you mean it "crashes" when it's not open...?

Does the phone show any sign of a crash, or it's just your logs that tell you?

**Comment by lucianoiam** [comment]
Android pops up a "XXXX keeps stopping" dialog in addition to the stack trace in logcat

**Comment by kari538** [comment]
Oh!... Yeah, definitely worse than I thought.

But if you have gotten it to work under other circumstances, perhaps you could help me with THIS issue: https://github.com/FirebaseExtended/flutterfire/issues/4887 ? Please? 😇 I get incompatible types in my Application.java file...

**Comment by kari538** [comment]
Btw, have you tried calling initializeApp() in your "onMessage" callback...? Making it async?

Just a thought.

**Comment by lucianoiam** [comment]
Please stop spamming this post and confusing other users.

- If you have a working solution please open a pull request
- If you need someone to work on your projects please check out UpWork

**Comment by lucianoiam** [comment]
> Btw, have you tried calling initializeApp() in your "onMessage" callback...? Making it async?
> 
> Just a thought.

That does not make sense


> Oh!... Yeah, definitely worse than I thought.
> 
> But if you have gotten it to work under other circumstances, perhaps you could help me with THIS issue: #4887 ? Please? 😇 I get incompatible types in my Application.java file...

That will definitely not work, check out https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects

**Comment by lucianoiam** [comment]
Here is a workaround that enables apps to continue receiving messages after being killed. As a workaround it might come with unknown side effects. The plugin integration docs are probably outdated https://pub.dev/documentation/firebase_messaging/latest/ . `PluginRegistrantCallback` is marked as deprecated.

```
import io.flutter.app.FlutterApplication;
import io.flutter.view.FlutterMain;

public class Application extends FlutterApplication {
    @Override
    public void onCreate() {
        super.onCreate();
        FlutterMain.ensureInitializationComplete(this, new String[]{});
    }

}
```

**Comment by tusharnikam-htp** [comment]
@lucianoiam Does this workaround applicable on Flutter [v2 embedding](https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects) also?

I don't think  `import io.flutter.app.FlutterApplication;` and `import io.flutter.view.FlutterMain;` applicable in V2 embedding.

**Comment by lucianoiam** [comment]
No idea

**Comment by tusharnikam-htp** [comment]
can you please reopen this issue then flutter team give a better solution

**Comment by lucianoiam** [comment]
As described above the plugin integration docs are outdated. That is probably the real issue. Opening a new issue stating that might attract better attention than this one which is rather generic.

**Comment by kari538** [comment]
@lucianoiam  could you plz provide a bit more info on your work-around? When I try to build my app with the Application.java file just the way you have written it, I get this error:

```
C:\Users\karol\AndroidStudioProjects\my_giggz\android\app\src\main\java\com\karolinadart\my_giggz\Application.java:12: error: cannot find symbol
        FlutterInjector.instance().flutterLoader().ensureInitializationComplete(this, new String[]{});
        ^
  symbol:   variable FlutterInjector
  location: class Application
1 error

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.
```

Android Studio suggests I change FlutterInjector to some class that exists in the `io.flutter.FlutterInjector` library. (The `io.flutter.view.FlutterMain;` import has all along been marked as "unused".) When I do that, the app builds and works like before... but I still can't receive any notifications when the app is terminated!

Btw, only once did I get the message "<app_name> keeps crashing!" like you said you got... The rest of the times, I just don't see anything on the phone, but the logs tell me a fatal error has occurred...
